### PR TITLE
SONARVB-238 Rule S1075: URIs should not be hardcoded

### PR DESCRIPTION
**Pull request by michalb-sonar**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        private const string PathDelimiterMessage = "Remove this hardcoded path-delimiter.";
+
+        // Simplified implementation of specification listed on
+        // https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
+        private const string UriScheme = "^[a-zA-Z][a-zA-Z\\+\\.\\-]+://.+";
+        private const string AbsoluteDiskUri = @"^[A-Za-z]:(/|\\)";
+        private const string AbsoluteMappedDiskUri = @"^\\\\\w[ \w\.]*";
+        private const string AbsoluteUnixUri = @"^(~\\|~/|/)\w";
+
+        private static readonly Regex UriRegex =
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR A static field in a generic type is not shared among instances of different close constructed types. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2743)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            new Regex($"{UriScheme}|{AbsoluteDiskUri}|{AbsoluteMappedDiskUri}|{AbsoluteUnixUri}",
+                RegexOptions.Compiled);
+
+        private static readonly Regex PathDelimiterRegex = new Regex(@"^(\\|/)$", RegexOptions.Compiled);
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR A static field in a generic type is not shared among instances of different close constructed types. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2743)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+        private static readonly ISet<string> checkedVariableNames = ImmutableHashSet.Create(
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR A static field in a generic type is not shared among instances of different close constructed types. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2743)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class UriShouldNotBeHardcodedBase<TExpressionSyntax,
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 5]
```diff
@@ -1,153 +1,55 @@
...
+        protected override ExpressionSyntax GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
+        protected override ExpressionSyntax GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
+        protected override bool IsInvocationOrObjectCreation(SyntaxNode node) =>
+            node.IsKind(SyntaxKind.InvocationExpression) ||
+            node.IsKind(SyntaxKind.ObjectCreationExpression);
+
+        protected override int? GetArgumentIndex(ArgumentSyntax argument) =>
+            (argument?.Parent as ArgumentListSyntax)?.Arguments.IndexOf(argument);
+
+        protected override string GetDeclaratorIdentifierName(VariableDeclaratorSyntax declarator)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: CRITICAL Rename parameter 'declarator' to 'literalExpression' to match the base class declaration. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS927)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        protected override ExpressionSyntax GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
+        protected override ExpressionSyntax GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
+        protected override bool IsInvocationOrObjectCreation(SyntaxNode node) =>
+            node.IsKind(SyntaxKind.InvocationExpression) ||
+            node.IsKind(SyntaxKind.ObjectCreationExpression);
+
+        protected override int? GetArgumentIndex(ArgumentSyntax argument)
+            => (argument?.Parent as ArgumentListSyntax)?.Arguments.IndexOf(argument);
+
+        protected override string GetDeclaratorIdentifierName(VariableDeclaratorSyntax declarator)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: CRITICAL Rename parameter 'declarator' to 'literalExpression' to match the base class declaration. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS927)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        protected static readonly ISet<string> checkedVariableNames = ImmutableHashSet.Create(
+            "file",
+            "path",
+            "uri",
+            "url",
+            "urn",
+            "stream");
+    }
+
+    public abstract class UriShouldNotBeHardcodedBase<TExpressionSyntax,
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs (hunk 7 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs (hunk 7 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class UriShouldNotBeHardcoded
+        : UriShouldNotBeHardcodedBase<ExpressionSyntax, LiteralExpressionSyntax,
+            SyntaxKind, BinaryExpressionSyntax, ArgumentSyntax, VariableDeclaratorSyntax>
+    {
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        protected override DiagnosticDescriptor Rule { get; } = rule;
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: I think if would be better if you use `protected override DiagnosticDescriptor Rule => rule;` instead.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        protected override SyntaxKind StringLiteralSyntaxKind { get; } = SyntaxKind.StringLiteralExpression;
+        protected override SyntaxKind[] StringConcatenateExpressions { get; } = new[] {
+            SyntaxKind.AddExpression,
+            SyntaxKind.ConcatenateExpression};
+        protected sealed override GeneratedCodeRecognizer GeneratedCodeRecognizer => Helpers.VisualBasic.GeneratedCodeRecognizer.Instance;
+        protected override string GetLiteralText(LiteralExpressionSyntax literalExpression) => literalExpression?.Token.ValueText;
+        protected override ExpressionSyntax GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
+        protected override ExpressionSyntax GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
+        protected override bool IsInvocationOrObjectCreation(SyntaxNode node) =>
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: I don't like the lack of whitespace in the wall of code above :)

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/DoNotUseByValTest.cs [hunk 10]
```diff
@@ -1,48 +1,48 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace SonarAnalyzer.UnitTest.Rules
+{
+    [TestClass]
+    public class DoNotUseByValTest
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/DoNotUseByValTest.cs]: Why is this file here?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 11]
```diff
@@ -1,153 +1,55 @@
...
+            => declarator?.Identifier.ValueText;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: all other arrows are on the previous line. Please move this one.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+using System.Collections.Immutable;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: System directives first please.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: Please remove this extra empty line.

File: sonaranalyzer-dotnet/its/expected/Ember-MM/Ember Media Manager-{9B57D3AB-AF12-4012-B945-284C2448DC81}-S1075.json [hunk 14]
```diff
@@ -0,0 +1,147 @@
+{
+"issues":  [
+{
+"id":  "S1075",
+"message":  "Refactor your code not to use hardcoded absolute paths or URIs.",
+"location":  {
+"uri":  "sources\Ember-MM\Ember%20Media%20Manager\dlgAbout.vb",
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/expected/Ember-MM/Ember Media Manager-{9B57D3AB-AF12-4012-B945-284C2448DC81}-S1075.json]: Pretty sure you checked them all but are they all real issues?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        : UriShouldNotBeHardcodedBase<ExpressionSyntax, LiteralExpressionSyntax,
+            SyntaxKind, BinaryExpressionSyntax, ArgumentSyntax, VariableDeclaratorSyntax>
+    {
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        protected override DiagnosticDescriptor Rule { get; } = rule;
+        protected override SyntaxKind StringLiteralSyntaxKind { get; } = SyntaxKind.StringLiteralExpression;
+        protected override SyntaxKind[] StringConcatenateExpressions { get; } = new[] {
+            SyntaxKind.AddExpression,
+            SyntaxKind.ConcatenateExpression};
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: Please add a space after `ConcatenateExpression`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        protected sealed override GeneratedCodeRecognizer GeneratedCodeRecognizer => Helpers.VisualBasic.GeneratedCodeRecognizer.Instance;
+        protected override string GetLiteralText(LiteralExpressionSyntax literalExpression) => literalExpression?.Token.ValueText;
+        protected override ExpressionSyntax GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
+        protected override ExpressionSyntax GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
+        protected override bool IsInvocationOrObjectCreation(SyntaxNode node) =>
+            node.IsKind(SyntaxKind.InvocationExpression) ||
+            node.IsKind(SyntaxKind.ObjectCreationExpression);
+
+        protected override int? GetArgumentIndex(ArgumentSyntax argument)
+            => (argument?.Parent as ArgumentListSyntax)?.Arguments.IndexOf(argument);
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: Please put the arrow on previous line.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+        protected override string GetDeclaratorIdentifierName(VariableDeclaratorSyntax declarator)
+            => declarator.Names.FirstOrDefault()?.Identifier.ValueText;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: Please put the arrow on previous line.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs (hunk 8 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: Oh, I missed that you can do it for properties. Good tip!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs (hunk 9 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: It looks quite nice in IDE, less so on GH.

File: sonaranalyzer-dotnet/its/expected/Ember-MM/Ember Media Manager-{9B57D3AB-AF12-4012-B945-284C2448DC81}-S1075.json (hunk 14 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/its/expected/Ember-MM/Ember Media Manager-{9B57D3AB-AF12-4012-B945-284C2448DC81}-S1075.json]: Yep.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/DoNotUseByValTest.cs (hunk 10 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Rules/DoNotUseByValTest.cs]: No idea, will revert.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs (hunk 17 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs]: We don't need stylecop with you around :)
Good comments, agree with all of them.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+        protected static readonly ISet<string> checkedVariableNames = ImmutableHashSet.Create(
+            "file",
+            "path",
+            "uri",
+            "url",
+            "urn",
+            "stream");
+    }
+
+    public abstract class UriShouldNotBeHardcodedBase<TExpressionSyntax,
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs (hunk 18 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs (hunk 18 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs]: MAJOR Reduce the number of generic parameters in the 'UriShouldNotBeHardcodedBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
